civis/io/_files.py

```diff
@@ -167,3 +180,154 @@ def _get_url_from_file_id(file_id, client):
+    # Retrieve run outputs
+    try:
+        outputs = client.scripts.list_containers_runs_outputs(job_id, run_id)
+    except CivisAPIError as err:
+        if err.status_code == 404:
+            raise IOError('Could not find job/run ID {}/{}'
+                          .format(job_id, run_id)) from err
+        else:
+            raise
+
+    # Find file in the run outputs.
+    if not regex:
+        # Require an exact match on the "name" string.
+        obj = find_one(outputs, name=name, object_type='File')
+    else:
+        # Search for a filename which contains the "name" string
+        obj_matches = [o for o in outputs
+                       if re.search(name, o.name) and o.object_type == 'File']
+        if len(obj_matches) > 1:
+            log.warning('Found %s matches to "%s". Returning the first.',
+                        len(obj_matches), name)
+        obj = None if not obj_matches else obj_matches[0]
+    if obj is None:
+        prefix = "A file containing the pattern" if regex else "File"
+        raise FileNotFoundError('{} "{}" is not an output of job/run ID '
+                                '{}/{}.'.format(prefix, name, job_id, run_id))
+    return obj['object_id']
+
+
+def file_to_dataframe(file_id, compression='infer', client=None,
+                      **read_kwargs):
+    """Load a :class:`~pandas.DataFrame` from a CSV stored in a Civis File
+
+    The :class:`~pandas.DataFrame` will be read directly from Civis
+    without copying the CSV to a local file on disk.
+
+    Parameters
+    ----------
+    file_id : int
+        ID of a Civis File which contains a CSV
+    compression : str, optional
+        If "infer", set the ``compression`` argument of ``pandas.read_csv``
+        based on the file extension of the name of the Civis File.
+        Otherwise pass this argument to ``pandas.read_csv``.
+    client : :class:`civis.APIClient`, optional
+        If not provided, an :class:`civis.APIClient` object will be
+        created from the :envvar:`CIVIS_API_KEY`.
+    **read_kwargs
+        Additional arguments will be passed directly to
+        :func:`~pandas.read_csv`.
+
+    Returns
+    -------
+    :class:`~pandas.DataFrame` containing the contents of the CSV
+
+    Raises
+    ------
+    ImportError
+        If ``pandas`` is not available
+
+    See Also
+    --------
+    :func:`~pandas.read_csv`
+    """
+    # Make sure we have pandas; this will raise an ImportError if not.
+    import pandas  # NOQA
+    client = APIClient() if client is None else client
+    file_info = client.files.get(file_id)
+    file_url = file_info.file_url
+    file_name = file_info.name
+    if compression == 'infer':
+        for ext in ['.bz2', '.zip', '.xz']:
+            if os.path.splitext(file_name)[-1] == ext:
+                compression = ext[1:]
+                break
+        else:
+            if os.path.splitext(file_name)[-1] == '.gz':
+                compression = 'gzip'
```

Comment:
I'd rather leave it as "infer" if we can't figure it out. In practice, I think we'll always end up with no compression in that case, but I believe pandas does check http headers to see if any compression format is indicated there.

Your code suggestion is still much clearer than what I wrote, so I changed to what you suggested (with an extra `if`).